### PR TITLE
Improve error handling of agent

### PR DESCRIPTION
**Pull request by nadheesh**
## Purpose
$subject

### CODE REVIEW

File: ballerina/agent.bal [hunk 1]
```diff
@@ -95,60 +94,58 @@ public class AgentExecutor {
...
-            return error LlmInputParseError("Error: Unable to extract the tool due to the invalid generation. Can you use the specified format?");
+            log:printWarn("Unexpected LLM response is given", llmResponse = llmResponse);
+            return error LlmActionParseError("Unable to extract the tool due to invalid generation", llmResponse = llmResponse, instruction = "Tool execution failed due to invalid generation. Regenerate following the given format.");
         }
-
         NextTool|error nextTool = content[1].fromJsonStringWithType();
         if nextTool is error {
-            log:printError(string `Error while extracting actions from LLM response. ${nextTool.toString()}`);
-            return error LlmInputParseError("Error: Provide with an invalid `action` JSON blob. Can you follow the specified format?");
+            log:printWarn("Unexpected JSON schema is given as the action.", nextTool);
```
Comment (JayaniH) [inline ballerina/agent.bal:111]: is it okay to make this a warning log when we are throwing an error after this?

File: ballerina/tool.bal [hunk 2]
```diff
@@ -55,13 +54,13 @@ isolated class ToolStore {
...
         if !self.tools.hasKey(toolName) {
-            return error ToolNotFoundError(string `Can't find the tool '${toolName}'. Provide a valid toolName`);
+            return error ToolNotFoundError("Cannot find the tool.", toolName = toolName, instruction = string `Tool "${toolName}" does not exists. Use a tool from the list: ${self.extractToolInfo().toolList}`);
         }
 
         isolated function caller = self.tools.get(toolName).caller;
         map<json>|error inputValues = mergeInputs(inputs, self.tools.get(toolName).constants);
         if inputValues is error {
-            return error ToolInvalidInputError(string `Tool '${toolName}' is provide with invalid inputs: ${(inputs ?: {}).toString()}`);
+            return error ToolInvalidInputError("Tool is provide with invalid inputs.", inputValues, toolName = toolName, inputs = inputs ?: {}, instruction = string `Tool "${toolName}"  execution failed due to invalid inputs provided. Use the schema to provide inputs: ${self.tools.get(toolName).variables.toString()}`);
```
Comment (JayaniH) [inline ballerina/tool.bal:63]:
```suggestion
            return error ToolInvalidInputError("Tool is provided with invalid inputs.", inputValues, toolName = toolName, inputs = inputs ?: {}, instruction = string `Tool "${toolName}"  execution failed due to invalid inputs. Use the schema to provide inputs: ${self.tools.get(toolName).variables.toString()}`);
```

Comment (JayaniH) [inline ballerina/tool.bal:63]: when we give the `inputValues` as it is, does it spread the values of the json as different key-value pairs?

File: ballerina/tool.bal [hunk 3]
```diff
@@ -72,8 +71,7 @@ isolated class ToolStore {
             observation = trap check function:call(caller, toolParams);
         }
         if observation is error && observation.message() == "{ballerina/lang.function}IncompatibleArguments" {
-            log:printWarn(string `Tool '${toolName}' is provide with invalid inputs: ${inputValues.toString()}`);
-            return error ToolInvalidInputError(string `Tool '${toolName}' is provide with invalid inputs: ${(inputs ?: {}).toString()}`);
+            return error ToolInvalidInputError("Tool is provide with invalid inputs.", observation, toolName = toolName, inputs = inputValues.length() == 0 ? {} : inputValues, instruction = string `Tool "${toolName}"  execution failed due to invalid inputs provided. Use the schema to provide inputs: ${self.tools.get(toolName).variables.toString()}`);
```
Comment (JayaniH) [inline ballerina/tool.bal:74]:
```suggestion
            return error ToolInvalidInputError("Tool is provided with invalid inputs.", observation, toolName = toolName, inputs = inputValues.length() == 0 ? {} : inputValues, instruction = string `Tool "${toolName}"  execution failed due to invalid inputs. Use the schema to provide inputs: ${self.tools.get(toolName).variables.toString()}`);
```

File: ballerina/agent.bal [hunk 4]
```diff
@@ -95,60 +94,58 @@ public class AgentExecutor {
...
                 tool: "complete",
                 isCompleted: true
             };
         }
-
         string[] content = regex:split(llmResponse + "<endtoken>", "```");
         if content.length() < 3 {
-            log:printError("Unexpected LLM response is given: \n`" + llmResponse + "`");
-            return error LlmInputParseError("Error: Unable to extract the tool due to the invalid generation. Can you use the specified format?");
+            log:printWarn("Unexpected LLM response is given", llmResponse = llmResponse);
```
[the remark this answers is not in the document]
Comment (nadheesh) [reply ballerina/agent.bal:106]: yes because we do not throw the error, it is usually handled by the Agent.

File: ballerina/agent.bal (hunk 1 above)
Comment (nadheesh) [reply ballerina/agent.bal:111]: yes because we do not throw the error, it is usually handled by the Agent.

File: ballerina/tool.bal (hunk 2 above)
Comment (nadheesh) [reply ballerina/tool.bal]: it will be a map<json> entry

**Review by JayaniH — APPROVED** [review]

**Review by malithjayasinghe — APPROVED** [review]

**Pull request merged**